### ISSUE TITLE
Reactive Tram Consumer

### ISSUE THREAD

**Issue by dartartem** [eventuate-tram/eventuate-tram-core#144]
### Goals & Background:

The goal of reactive processing is handle data when it is ready, without waiting for readiness.
In other words do not block processing (thread) for waiting IO operations.

There are 2 ways to make eventuate tram consumer reactive.

1) Replace kafka consumer by reactive version. https://projectreactor.io/docs/kafka/release/reference/

**Advantages (according to docs):**

- Functional style
- Automatic back-pressure handling

**Disadvantages:**

- Reactive API requieres completely rewrite eventuate consumers in eventuate-messaging-kafka project. It is large codebase.

**Thoughts:**

We already have [back-pressure handling](https://github.com/eventuate-foundation/eventuate-messaging-kafka/tree/master/eventuate-messaging-kafka-basic-consumer/src/main/java/io/eventuate/messaging/kafka/basic/consumer) and functional style is not important for internal details of eventuate kafka consumer (eventuate-messaging-kafka project)

2) Adapt existing API for reactive processing. What do we have now:

[MessageConsumerKafkaImpl](https://github.com/eventuate-foundation/eventuate-messaging-kafka/blob/master/eventuate-messaging-kafka-consumer/src/main/java/io/eventuate/messaging/kafka/consumer/MessageConsumerKafkaImpl.java) uses simple java consumer for event handling:

`public KafkaSubscription subscribe(String subscriberId, Set<String> channels, KafkaMessageHandler handler) {`

[link](https://github.com/eventuate-foundation/eventuate-messaging-kafka/blob/master/eventuate-messaging-kafka-consumer/src/main/java/io/eventuate/messaging/kafka/consumer/MessageConsumerKafkaImpl.java#L38)

```
public interface KafkaMessageHandler extends Consumer<KafkaMessage> {
}
```

[link](https://github.com/eventuate-foundation/eventuate-messaging-kafka/blob/a8d2c997621833307e735cce0b6216d10901094b/eventuate-messaging-kafka-consumer/src/main/java/io/eventuate/messaging/kafka/consumer/KafkaMessageHandler.java#L5)

Internally is wrapped up by [EventuateKafkaConsumerMessageHandler](https://github.com/eventuate-foundation/eventuate-messaging-kafka/blob/a8d2c997621833307e735cce0b6216d10901094b/eventuate-messaging-kafka-basic-consumer/src/main/java/io/eventuate/messaging/kafka/basic/consumer/EventuateKafkaConsumerMessageHandler.java)

```
public interface EventuateKafkaConsumerMessageHandler
        extends BiFunction<ConsumerRecord<String, byte[]>, BiConsumer<Void, Throwable>, MessageConsumerBacklog> {
}
```

```
EventuateKafkaConsumerMessageHandler kcHandler = (record, callback) ->
        swimlaneBasedDispatcher.dispatch(new RawKafkaMessage(record.value()),
                record.partition(),
                message -> handle(message, callback, handler));
```
```

  public void handle(RawKafkaMessage message, BiConsumer<Void, Throwable> callback, KafkaMessageHandler kafkaMessageHandler) {
    try {
      if (eventuateKafkaMultiMessageConverter.isMultiMessage(message.getPayload())) {
        eventuateKafkaMultiMessageConverter
                .convertBytesToMessages(message.getPayload())
                .getMessages()
                .stream()
                .map(EventuateKafkaMultiMessage::getValue)
                .map(KafkaMessage::new)
                .forEach(kafkaMessageHandler);
      } else {
        kafkaMessageHandler.accept(new KafkaMessage(EventuateBinaryMessageEncoding.bytesToString(message.getPayload())));
      }
      callback.accept(null, null);
    } catch (Throwable e) {
      callback.accept(null, e);
      throw e;
    }
  }
```

handler has callbacks to notify [message processor](https://github.com/eventuate-foundation/eventuate-messaging-kafka/blob/a8d2c997621833307e735cce0b6216d10901094b/eventuate-messaging-kafka-basic-consumer/src/main/java/io/eventuate/messaging/kafka/basic/consumer/KafkaMessageProcessor.java) that message handled:

**Comment by cer** [comment]
Here is a suggestion:

* There are functions that take `BiConsumer<Void, Throwable> callback` as an argument
   * callback is invoked with either `(ignored, null)` on success or `(null, throwable)` on failure
  * See above
  * [EventuateKafkaConsumerMessageHandler](https://github.com/eventuate-foundation/eventuate-messaging-kafka/blob/b7b618dff09401cf12a51b7b59dd4de0d2771e2d/eventuate-messaging-kafka-basic-consumer/src/main/java/io/eventuate/messaging/kafka/basic/consumer/EventuateKafkaConsumerMessageHandler.java#L8)

A Java CompletableFuture is something that contains a value or a Throwable.

Therefore:

* Instead of having a `BiConsumer<Void, Throwable> callback` as a parameter. These functions should return a CompletableFuture.
* In other words, the Kafka consumer becomes `reactive` in a standard Java way
* The Mono-based API can convert the Mono into a CompletableFuture.
*

**Comment by dartartem** [comment]
@cer yes, I planned to propose something similar, I just accidentally created partially filed issue.

**Comment by cer** [comment]
Also, I think it's important to define the goal here:

```
public interface ReactiveMessageConsumer {
  MessageSubscription subscribe(String subscriberId, Set<String> channels, ReactiveMessageHandler handler);
  String getId();
  void close();
}

public interface ReactiveMessageHandler extends Function<Message, Mono<Void> {
}
```

Reactive event handling:

```
  public DomainEventDispatcher(String eventDispatcherId, ReactiveDomainEventHandlers domainEventHandlers, MessageConsumer messageConsumer, DomainEventNameMapping domainEventNameMapping) {
```

```
 public ReactiveDomainEventHandlers domainEventHandlers() {
    return ReactiveDomainEventHandlersBuilder
            .forAggregateType("io.eventuate.examples.tram.ordersandcustomers.customers.domain.Customer")
            .onEvent(CustomerCreditReservedEvent.class, this::handleCustomerCreditReservedEvent)
            ...
            .build();
  }

  private Mono<Void> handleCustomerCreditReservedEvent(DomainEventEnvelope<CustomerCreditReservedEvent> domainEventEnvelope) {
    ...
  }
```
